core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -169,8 +169,10 @@ abstract class AbstractFetcherThread(name: String,
           val topic = topicPartition.topic
           val partitionId = topicPartition.partition
           Option(partitionStates.stateValue(topicPartition)).foreach(currentPartitionFetchState =>
-            // we append to the log if the current offset is defined and it is the same as the offset requested during fetch
-            if (fetchRequest.offset(topicPartition) == currentPartitionFetchState.fetchOffset) {
+            // We append to the log if (1) the state of the partition is ready for fetching, and (2) the current offset
+            // is defined and it is the same as the offset requested during fetch
+            if (currentPartitionFetchState.isReadyForFetch &&
```

Comment:
I thought `isReadyForFetch` was only used to decide if the partition was ready for fetching. Like in the following place:

```
partitionMap.foreach { case (topicPartition, partitionFetchState) =>
      // We will not include a replica in the fetch request if it should be throttled.
      if (partitionFetchState.isReadyForFetch && !shouldFollowerThrottle(quota, topicPartition)) {
        val logStartOffset = replicaMgr.getReplicaOrException(topicPartition).logStartOffset
        requestMap.put(topicPartition, new JFetchRequest.PartitionData(partitionFetchState.fetchOffset, logStartOffset, fetchSize))
      }
    }
```

What is the reason for not appending to the log in this case?

Reply:
This is to handle the case that a partition is removed and re-added quickly. In this case, the newly added partition needs to go through LeaderEpochRequest, log truncation, etc before it can fetch data again. So, a fetch request from the previous generation needs to be ignored.

Reply:
Thanks for the explanation. A few questions/comments:
1. Is it worth adding your explanation to the code since it doesn't seem obvious?
2. Maybe we want to introduce another method `isReadyForLogAppend` in `PartitionFetchState`.
3. Would it be possible to add a test for this fix?